From 8e9882d8910edf3f0ec51303864d023f152fa4e9 Mon Sep 17 00:00:00 2001
From: Ivan Kardykov <kardykov@tabit.pro>
Date: Thu, 9 Dec 2021 13:27:27 +0300
Subject: [PATCH] relocate-dir: support IO_REPARSE_TAG_APPEXECLINK

This fixes move-profile feature fails on Windows 10

This fixes QubesOS/qubes-issues#5093
---
 src/relocate-dir/io.c | 11 +++++++++++
 src/relocate-dir/io.h |  5 +++++
 2 files changed, 16 insertions(+)

diff --git a/src/relocate-dir/io.c b/src/relocate-dir/io.c
index d7d4c45..be95d3f 100644
--- a/src/relocate-dir/io.c
+++ b/src/relocate-dir/io.c
@@ -57,6 +57,8 @@ NTSTATUS FileOpen(OUT HANDLE *file, IN const PWCHAR fileName, IN BOOLEAN write,
     {
         if (!isReparse)
             desiredAccess |= FILE_WRITE_DATA | FILE_APPEND_DATA | FILE_WRITE_ATTRIBUTES | FILE_WRITE_EA | WRITE_DAC | WRITE_OWNER | DELETE;
+	else
+	    desiredAccess |= DELETE;
 
         if (overwrite)
         {
@@ -528,6 +530,7 @@ NTSTATUS FileDelete(IN HANDLE file)
         if (!NT_SUCCESS(status))
         {
             NtLog(TRUE, L"[!] FileDelete: FSCTL_DELETE_REPARSE_POINT failed: %x\n", status);
+            if (rdb->ReparseTag != IO_REPARSE_TAG_APPEXECLINK)
             goto cleanup;
         }
     }
@@ -700,6 +703,10 @@ NTSTATUS FileCopyReparsePoint(IN const PWCHAR sourcePath, IN const PWCHAR target
         NtLog(FALSE, L"[*] Mount point: '%s' -> '%s'\n", sourcePath, dest);
         break;
 
+    case IO_REPARSE_TAG_APPEXECLINK:
+	NtLog(FALSE, L"[*] AppExecLink: '%s'\n", sourcePath);
+	break;
+
     default:
 
         NtLog(TRUE, L"[!] FileCopyReparsePoint: Unknown reparse tag 0x%x in '%s'\n", rdb->ReparseTag, sourcePath);
@@ -1061,6 +1068,10 @@ NTSTATUS FileDeleteDirectory(IN const PWCHAR path, IN BOOLEAN deleteSelf)
                 {
                     // just delete
                     status = FileOpen(&file, fullPath, TRUE, FALSE, FALSE);
+                    
+                    if ((!NT_SUCCESS(status)) && (entry->FileAttributes  & FILE_ATTRIBUTE_REPARSE_POINT))
+		        status = FileOpen(&file, fullPath, TRUE, FALSE, TRUE);
+
                     if (!NT_SUCCESS(status))
                     {
                         NtLog(TRUE, L"[!] FileOpen(%s) failed: %x\n", fullPath, status);
diff --git a/src/relocate-dir/io.h b/src/relocate-dir/io.h
index c834985..95701a9 100644
--- a/src/relocate-dir/io.h
+++ b/src/relocate-dir/io.h
@@ -50,3 +50,8 @@ NTSTATUS FileCopyReparsePoint(IN const PWCHAR sourcePath, IN const PWCHAR target
 NTSTATUS FileSetSymlink(IN const PWCHAR sourcePath, IN const PWCHAR targetPath);
 NTSTATUS FileCopyDirectory(IN const PWCHAR sourcePath, IN const PWCHAR targetPath, IN BOOLEAN ignoreErrors);
 NTSTATUS FileDeleteDirectory(IN const PWCHAR path, IN BOOLEAN deleteSelf);
+
+#ifndef IO_REPARSE_TAG_APPEXECLINK
+#define IO_REPARSE_TAG_APPEXECLINK (0x8000001BL)
+#endif
+
-- 
2.31.1

